sdk/core/azure-core/azure/core/rest/_http_response_impl.py

```diff
@@ -0,0 +1,208 @@
+# --------------------------------------------------------------------------
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
+#
+# The MIT License (MIT)
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the ""Software""), to
+# deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+from json import loads
+from typing import cast, Any, Optional, Iterator
+from ._helpers import (
+    get_charset_encoding,
+    decode_to_text,
+    parse_lines_from_text,
+)
+from ..exceptions import HttpResponseError, ResponseNotReadError
+try:
+    from ._rest_py3 import (
+        _HttpResponseBase,
+        HttpResponse as _HttpResponse,
+        HttpRequest as _HttpRequest
+    )
+except (SyntaxError, ImportError):
+    from ._rest import (  # type: ignore
+        _HttpResponseBase,
+        HttpResponse as _HttpResponse,
+        HttpRequest as _HttpRequest
+    )
+
+
+class _HttpResponseBaseImpl(_HttpResponseBase):
+
+    def __init__(self, **kwargs):
+        # type: (Any) -> None
+        super(_HttpResponseBaseImpl, self).__init__()
+        self._request = kwargs.pop("request")
+        self._internal_response = kwargs.pop("internal_response")
+        self._is_closed = False
+        self._is_stream_consumed = False
+        self._connection_data_block_size = None
+        self._json = None  # this is filled in ContentDecodePolicy, when we deserialize
+        self._content = None  # type: Optional[bytes]
+        self._text = None  # type: Optional[str]
+
+    @property
+    def request(self):
+        # type: (...) -> _HttpRequest
```

Comment:
This should return an `azure.core.rest.HttpRequest`, correct?

Reply:
Yeah, the typing's correct, let me add a docstring there though, that one did get missed